### PR TITLE
bpo-24959: fix unittest.assertRaises bug where traceback entries are dropped from chained exceptions

### PR DESCRIPTION
**Pull request by iritkatriel**
Unittest applies a hack that uses the limit parameter of TracebackException to hide traceback frames from the unittest code itself. The problem with this is that the limit applies to the chained (context) exception too, so we lose frames we want to keep.

This patch replaces the hack by code that modifies the traceback to remove the frames we don't want.

https://bugs.python.org/issue24959

### CODE REVIEW

**Comment by iritkatriel** [comment]
https://bugs.python.org/issue42247

**Comment by iritkatriel** [comment]
See also https://bugs.python.org/issue24959

File: Lib/unittest/test/test_result.py [hunk 1]
```diff
@@ -413,6 +413,47 @@ class tb_frame(object):
         Frame.tb_frame.f_globals['__unittest'] = True
         self.assertTrue(result._is_relevant_tb_level(Frame))
 
+    def testRemoveUnittestTbFrames(self):
+        try:
+            self.fail('too bad')
+        except:
```
Comment (serhiy-storchaka) [inline Lib/unittest/test/test_result.py]:
```suggestion
        except self.failureException:
```
We do not want to swallow KeyboardInterrupt etc.

File: Lib/unittest/test/test_result.py [hunk 2]
```diff
@@ -413,6 +413,47 @@ class tb_frame(object):
...
+        result._remove_unittest_tb_frames(tb)
+        self.assertEqual(len(list(traceback.walk_tb(tb))), 1)
+
+    def testExcInfoStringChained(self):  # bpo 49563
+        def raiseAnException():
+            raise ValueError(42)
+        try:
+            try:
+                raiseAnException()
+            except:
```
Comment (serhiy-storchaka) [inline Lib/unittest/test/test_result.py]:
```suggestion
            except ValueError:
```

File: Lib/unittest/test/test_result.py [hunk 3]
```diff
@@ -413,6 +413,47 @@ class tb_frame(object):
...
+                self.fail('too bad')
+        except:
```
Comment (serhiy-storchaka) [inline Lib/unittest/test/test_result.py]:
```suggestion
        except self.failureException:
```

**Review by serhiy-storchaka — COMMENTED** [review]
LGTM in general.

I think that tracebacks in `__context__` (and maybe `__cause__`) chains should be cleaned recursively too. For the case like:
```python
try:
    self.assertFoo()
finally:
    self.assertBar()
```

**Comment by serhiy-storchaka** [comment]
Is it possible to write tests which do not rely on private methods?

**Comment by iritkatriel** [comment]
Thanks.   I've marked this as draft because I wanted consider a different approach to this.

We recently had some changes in the traceback module that allow you do customize how frames are rendered, and also to exclude frames from the output. I think it might be better to use that approach here.  I'll make another PR soon.

**Comment by iritkatriel** [comment]
I rewrote the test to not rely on private methods. 

The traceback utility we added recently doesn't help because it allows to filter FrameSummary objects (not frames) so we no longer have access to the globals.  

Regarding filtering chained tracebacks - I think this is another problem. The current filtering assumes that the traceback starts in user code. With chained tracebacks I'm not sure it is (in the contrived test I tried this wasn't the case).   I think we should define the use case and desired behaviour before implementing.

**Comment by iritkatriel** [comment]
Ah I've rebased against main as well.

**Comment by serhiy-storchaka** [comment]
The use case is:
```
prepare()
try:
    test()
finally:
    cleanup()
```
We want to acquire some resource or make some environment changes for the part of the test and release the resource or restore environment after testing, so we use the `try/finally` block. Now, if an assertion failed inside the block, and an error occurred in the clean up code, the new exception will override the original AssertionError which is now accessible via the `__context__` link. The traceback output will contain lines from the unittest code. Note that tracebecks for `__context__` should be cleaned up not only if `exctype is test.failureException`, but if `type(value.__context__) is test.failureException`,  `type(value.__context__.__context__) is test.failureException`, etc. But it may be a different issue.

I do not know the use case for `__cause__`, but I would clean up them too.

**Review by serhiy-storchaka — COMMENTED** [review]
LGTM, but I'll try to write more tests. Surprisingly, there are no tests for removing unittest frames at all.

**Comment by iritkatriel** [comment]
See testStackFrameTrimming in test_result.

**Comment by iritkatriel** [comment]
I added the chained exceptions. I avoided recursion because it can be just a bit too deep for certain constructs involving recursion.

**Comment by yashk2810** [comment]
Hello,

Any updates on the PR? Is it possible to get this in soon? :)

**Review by serhiy-storchaka — APPROVED** [review]
LGTM!

**Comment by iritkatriel** [comment]
Thank you @serhiy-storchaka !

**Comment by yashk2810** [comment]
Just to clarify: Will this work with python 3.7 or it would need to be backported?

**Comment by iritkatriel** [comment]
It will not be backport Ed to 3.7.
https://devguide.python.org/devcycle/

**Pull request merged**